Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -211,6 +211,54 @@ bool TextureCache::TCacheEntryBase::OverlapsMemoryRange(u32 range_address, u32 r
 	return true;
 }
 
+void TextureCache::TCacheEntryBase::DoPartialTextureUpdates()
+{
+	const bool isPaletteTexture = (format== GX_TF_C4 || format == GX_TF_C8 || format == GX_TF_C14X2 || format >= 0x10000);
+
+	// As long as there's no upscaling support, there's no point in doing this when doing efb to ram
+	// TODO: Only update normal textures. Would there be any point in updating efb copies?
+	// TODO: Are texture updates for paletted textures possible?
+	// Efb copies might be scaled, while normal textures are not. So don't try to update normal textures with scaled efb copies
+	if (!g_ActiveConfig.backend_info.bSupportsCopySubImage || !g_ActiveConfig.bSkipEFBCopyToRam || IsEfbCopy() || isPaletteTexture
+		|| (g_ActiveConfig.bCopyEFBScaled && g_ActiveConfig.iEFBScale != SCALE_1X))
+		return;
+
+	u32 block_width = TexDecoder_GetBlockWidthInTexels(format);
+	u32 block_height = TexDecoder_GetBlockHeightInTexels(format);
+	u32 block_size = block_width * block_height * TexDecoder_GetTexelSizeInNibbles(format) / 2;
+
+	u32 expanded_width = (native_width + block_width - 1) & ~(block_width - 1);
+
+	u32 numBlocksX = expanded_width / block_width;
+
+	TexCache::iterator iter = textures_by_address.lower_bound(addr);
+	TexCache::iterator iterend = textures_by_address.upper_bound(addr + size_in_bytes);
+
+	while (iter != iterend)
+	{
+		TCacheEntryBase* entry = iter->second;
+		if (entry->IsEfbCopy() && entry->frameCount == FRAMECOUNT_INVALID && entry->copyMipMapStrideChannels * 32 == numBlocksX * block_size
+			&& addr <= entry->addr && entry->addr + entry->size_in_bytes <= addr + size_in_bytes
+			&& !(native_width == entry->native_width && native_height == entry->native_height))
+		{
+			u32 block_offset = (entry->addr - addr) / block_size;
+			u32 block_x = block_offset % numBlocksX;
+			u32 block_y = block_offset / numBlocksX;
+
+			u32 x = block_x * block_width;
+			u32 y = block_y * block_height;
+
+			ERROR_LOG(VIDEO, "Doing partial texture update for a texture at: 0x%08x, efb copy at: 0x%08x, w/h: %4d/%4d, update at: x/y: %4d/%4d, w/h: %4d/%4d", addr, entry->addr, native_width, native_height, x, y, entry->native_width, entry->native_height);
```

Comment:
It will be removed before merging. With logging enabled, the 1000s of messages kill the performane of the emulator. Like after playing 1-2 levels in New Super Mario Bros, the fps went down to 40-50 for me, with framelimt auto. While holding hold tab in that situation, it only got to 200 fps, so i think there's some issue with the log. It was even only set to log to window, and the window was not even active.